client/engine/engine.go

```diff
@@ -0,0 +1,135 @@
+type Response struct{}
+
+// NewEngine is the constructor for an Engine
+func New() Engine {
+	e := Engine{}
+
+	// create the engine's inbound channels
+	e.API = make(chan APIEvent)
+	e.Chain = make(chan ChainEvent)
+	e.Inbox = make(chan Message)
+
+	return e
+}
+
+// Run kicks of an infinite loop that waits for communications on the supplied channels, and handles them accordingly
+func (e *Engine) Run() {
+	for {
+		select {
+		case apiEvent := <-e.API:
+			e.handleAPIEvent(apiEvent)
+
+		case chainEvent := <-e.Chain:
+			e.handleChainEvent(chainEvent)
+
+		case message := <-e.Inbox:
+			e.handleMessage(message)
+
+		}
+	}
+}
+
+// handleMessage handles a Message from a peer go-nitro Wallet.
+// It
+// reads an objective from the store,
+// generates an updated objective and declaration of side effects,
+// commits the updated objective to the store,
+// executes the side effects and
+// evaluates objecive progress.
+func (e *Engine) handleMessage(message Message) {
+	objective := e.store.GetObjectiveById(message.ObjectiveId)
+	event := protocols.ObjectiveEvent{Sigs: message.Sigs}
+	updatedProtocol := objective.Update(event)
+	e.store.SetObjective(updatedProtocol)
+	sideEffects, waitingFor, _ := updatedProtocol.Crank() // TODO handle error
+	e.executeSideEffects(sideEffects)
+	e.store.EvaluateProgress(message.ObjectiveId, waitingFor)
+}
+
+// handleChainEvent handles a Chain Event from the blockchain.
+// It
+// reads an objective from the store,
+// generates an updated objective and declaration of side effects,
+// commits the updated objective to the store,
+// executes the side effects and
+// evaluates objecive progress.
+func (e *Engine) handleChainEvent(chainEvent ChainEvent) {
+	objective := e.store.GetObjectiveByChannelId(chainEvent.ChannelId)
+	event := protocols.ObjectiveEvent{Holdings: chainEvent.Holdings, AdjudicationStatus: chainEvent.AdjudicationStatus}
+	updatedProtocol := objective.Update(event)
+	e.store.SetObjective(updatedProtocol)
+	sideEffects, waitingFor, _ := updatedProtocol.Crank() // TODO handle error
+	e.executeSideEffects(sideEffects)
+	e.store.EvaluateProgress(objective.Id(), waitingFor)
+
+}
+
+// handleAPIEvent handles an API Event (triggered by an API call)
+// It will perform one of the following, in priority order:
+// Spawn a new, approved objective
+// Reject an existing objective
+// Approve an existing objective
+func (e *Engine) handleAPIEvent(apiEvent APIEvent) {
+	switch {
+	case apiEvent.ObjectiveToSpawn != nil:
+		e.store.SetObjective(apiEvent.ObjectiveToSpawn)
+	case apiEvent.ObjectiveToReject != ``:
+		objective := e.store.GetObjectiveById(apiEvent.ObjectiveToReject)
+		updatedProtocol := objective.Reject()
+		e.store.SetObjective(updatedProtocol)
+	case apiEvent.ObjectiveToApprove != ``:
+		objective := e.store.GetObjectiveById(apiEvent.ObjectiveToReject)
+		updatedProtocol := objective.Approve()
+		e.store.SetObjective(updatedProtocol)
```

Comment:
That's an interesting suggestion. I think there are a couple of aspects to this:

1. Spawning objectives. At some point we need to put an objective in the store -- so I think the "core pattern" you describe can't apply.
2. Approving / Rejecting objectives. This could follow the "core pattern". But the core pattern favours expanding the set of events and reducing the size of the `Objective` interface. As we have discussed elsewhere, it is not straightforward to have a discrminated union type and a handler for that type which can be statically checked to handle all of the types in the union. So adding an event to the union may increase the chance that the `reject` event is not handled (for example). By having the methods in the interface of an `Objective`, we _can_ benefit from static checks and be sure that approvals / rejections are handled. I think it also increases the modularity of objective implementations (i.e. several smaller functions rather than one larger one).